compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs

```diff
@@ -1835,6 +1836,58 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         }
     }
 
+    pub(super) fn collect_unused_for_coerce_return_ty(
+        &self,
+        errors: &[traits::FulfillmentError<'tcx>],
+    ) {
+        for fillment_error in &*errors {
+            let obligation = &fillment_error.obligation;
+            let span = obligation.cause.span;
+
+            let Some(mut diag) =
+                self.tcx.sess.diagnostic().steal_diagnostic(span, StashKey::MaybeForgetReturn)
+            else {
+                continue;
+            };
+
+            let root_obligation = &fillment_error.root_obligation;
+            if let Some(fn_sig) = self.body_fn_sig()
+                && let ExprBindingObligation(_, _, hir_id, ..) = root_obligation.cause.code()
+                && !fn_sig.output().is_unit() {
+                    let mut block_num = 0;
+                    let mut found_semi = false;
+                    for (_, node) in self.tcx.hir().parent_iter(*hir_id) {
+                        match node {
+                            hir::Node::Stmt(stmt) => if let hir::StmtKind::Semi(ref expr) = stmt.kind {
+                                let expr_ty = self.typeck_results.borrow().expr_ty(expr);
+                                let return_ty = fn_sig.output();
+                                if !matches!(expr.kind, hir::ExprKind::Ret(..)) &&
+                                    self.can_coerce(expr_ty, return_ty) {
+                                    found_semi = true;
+                                }
+                            },
+                            hir::Node::Block(_block) => if found_semi {
+                                block_num += 1;
+                            }
+                            hir::Node::Item(item) => if let hir::ItemKind::Fn(..) = item.kind {
+                                break;
+                            }
+                            _ => {}
+                        }
+                    }
+                    if block_num > 1 && found_semi {
+                        diag.span_suggestion_verbose(
+                            span.shrink_to_lo(),
+                            "consider adding `return` here to coerce to expected return type",
```

Comment:
My only concern that remains is the wording here...

I kinda want this to say something like "did you forget to return this?" or something, to emphasize the fact the reason that the reason there was an ambiguity error here was because it was not equated to the return type.

Reply:
I would change this to be something like "you might have meant to return this, doing so would allow `rustc` to infer its type parameters".

Reply:
change it to `you might have meant to return this to infer its type parameters`.